### ISSUE TITLE
Adjoint for `weights`

### ISSUE THREAD

**Issue by vboussange** [gdalle/GridGraphs.jl#16]
I would be interested in being able to do something like this:

```Julia
using GridGraphs
using DifferentiationInterface
using Zygote

function objective(permeability)
    grid = GridGraph(permeability, directions=ROOK_DIRECTIONS)
    A = weights(grid)
    sum(A)
end

len = 100 # height and width of the grid
permeability = rand(Float32, len, len)
@time val, grad = value_and_gradient(objective, AutoZygote(), permeability)
```
But this fails
`ERROR: Need an adjoint for constructor SparseMatrixCSC{Float32, Int64}. Gradient is of type SparseMatrixCSC{Float32, Int64}`.

Any idea on how to do that?
Cheers!

**Comment by vboussange** [comment]
I agree with you @gdalle that diagonal moves may be an overkill :). But we could also think of more generality on how the user wants to define the edge weights from the vertices values. For instance, one may want to define the weight of an edge as the mean of the source and target vertex value.

Here is a non `GridGraphs` implementation that has this flexibility, but that still builds the weight matrix one coefficient at a time:
```julia
const N4 = (( 0, -1, 1.0),
            (-1,  0, 1.0),
            ( 1,  0, 1.0),
            ( 0,  1, 1.0))
const N8 = ((-1, -1,  √2),
            ( 0, -1, 1.0),
            ( 1, -1,  √2),
            (-1,  0, 1.0),
            ( 1,  0, 1.0),
            (-1,  1,  √2),
            ( 0,  1, 1.0),
            ( 1,  1,  √2))

"""
    graph_matrix_from_raster(R::Matrix[, type=AffinityMatrix, neighbors::Tuple=N8, weight=TargetWeight])::SparseMatrixCSC

Compute a graph matrix, i.e. an affinity or cost matrix of the raster image `R` of cell affinities or cell costs.
The values are computed as either the value of the target cell (TargetWeight) or as harmonic (arithmetic) means
of the cell affinities (costs) weighted by the grid distance (AverageWeight). The values can be computed with
respect to eight neighbors (`N8`) or four neighbors (`N4`).
"""
function graph_matrix_from_raster(
    R::Matrix;
    matrix_type=AffinityMatrix,
    neighbors::Tuple=N8,
    weight=TargetWeight
)
    m, n = size(R)

    # Initialize the buffers of the SparseMatrixCSC
    is, js, vs = Int[], Int[], eltype(R)[]

    for j in 1:n
        for i in 1:m
            # Base node
            rij = R[i, j]
            for (ki, kj, l) in neighbors
                if !(1 <= i + ki <= m) || !(1 <= j + kj <= n)
                    # Continue when computing edge out of raster image
                    continue
                else
                    # Target node
                    rijk = R[i + ki, j + kj]
                    if iszero(rijk) || isnan(rijk)
                        # Don't include zero or NaN similaritiers
                        continue
                    end

                    push!(is, (j - 1)*m + i)
                    push!(js, (j - 1)*m + i + ki + kj*m)
                    if weight == TargetWeight
                        if matrix_type == AffinityMatrix
                            push!(vs, rijk/l)
                        elseif matrix_type == CostMatrix
                            push!(vs, rijk*l)
                        end
                    elseif weight == AverageWeight
                        if matrix_type == AffinityMatrix
                            v = 2/((inv(rij) + inv(rijk))*l)
                            push!(vs, v)
                        elseif matrix_type == CostMatrix
                            v = ((rij + rijk)*l)/2
                            push!(vs, v)
                        end
                    else
                        throw(ArgumentError("weight mode $weight not implemented"))
                    end
                end
            end
        end
    end
    return sparse(is, js, vs, m*n, m*n)
end
```

I have not tried yet differentiating your `weights` function with Enzyme, but anyhow we could surely benefit from building the adjacency matrix in one go! Any idea on how to do it?

**Comment by gdalle** [comment]
> I have not tried yet differentiating your weights function with Enzyme

I tried it with your code from #15 and it threw an error, I'll investigate.

> anyhow we could surely benefit from building the adjacency matrix in one go! Any idea on how to do it?

I was thinking it might be possible to figure out an explicit formula for the `k`-th coefficient of `I`, `J` and `V` in the constructor `sparse(I, J, V)` if we restrict ourselves to lateral and vertical moves. But it's very tedious.

**Comment by vboussange** [comment]
> I tried it with your code from https://github.com/gdalle/GridGraphs.jl/issues/15 and it threw an error, I'll investigate.

Yep. For this reason, and a few other things that were bugging me, I gave it a try with JAX, and found a working solution - see below. The pseudo algorithm should be compatible with Zygote. When I find time, I could port this to Julia and make a PR.

```python

def build_adjacency_matrix(gridgraph):
    """
    Create a differentiable adjacency matrix based on the vertices and
    contiguity pattern of `gridgraph`.
    """
    assert isinstance(gridgraph, GridGraph)
    # Get shape of raster
    activities = gridgraph.activities
    nrows, ncols = activities.shape
    num_nodes = gridgraph.nb_active()
    neighbors = gridgraph.neighbors
    permeability_raster = gridgraph.vertex_weights
    
    source_xy_coord = gridgraph.active_vertex_index_to_coord(jnp.arange(num_nodes))
    active_map = jnp.zeros_like(activities, dtype=int) - 1
    active_map = active_map.at[source_xy_coord[:,0], source_xy_coord[:,1]].set(jnp.arange(num_nodes))  # -1 if not an active vertex
    
    rows, cols, values = [], [], []
    
    for neighb in neighbors:
        candidate_target_xy_coord = source_xy_coord + neighb
        
        # filter edges with out out-of-bound indices
        edge_validity = (candidate_target_xy_coord[:, 0] >= 0) & (candidate_target_xy_coord[:, 0] < nrows) & (candidate_target_xy_coord[:, 1] >= 0) & (candidate_target_xy_coord[:, 1] < ncols)

        # filter out edges with inactive target pixels
        edge_validity = edge_validity * activities[candidate_target_xy_coord[:,0], candidate_target_xy_coord[:,1]]
        
        source_node_coord = active_map[source_xy_coord[:, 0][edge_validity], source_xy_coord[:, 1][edge_validity]]
        target_node_coord = active_map[candidate_target_xy_coord[:, 0][edge_validity], candidate_target_xy_coord[:, 1][edge_validity]]
        
        # Connectivity values based on permeability
        value = permeability_raster[candidate_target_xy_coord[:, 0][edge_validity], candidate_target_xy_coord[:, 1][edge_validity]]

        values.extend(value)
        rows.extend(source_node_coord)
        cols.extend(target_node_coord)

    # Stack results into arrays for COO format
    data = jnp.array(values)
    row_col_indices = jnp.vstack([jnp.array(rows), jnp.array(cols)]).T

    A = BCOO((data, row_col_indices), shape=(num_nodes, num_nodes))

    return A
```

**Comment by gdalle** [comment]
I admire the effort, but this is a lot of work just to ensure compatibility with Zygote. Since Enzyme is not happy for some reason, a good alternative could be the brand new Mooncake, which also supports mutation. I tried it and it works like a charm:

```julia
using GridGraphs
using GridGraphs.Graphs
using DifferentiationInterface
import Mooncake

function objective(permeability)
    grid = GridGraph(permeability, directions=GridGraphs.ROOK_DIRECTIONS)
    A = weights(grid)
    return sum(A)
end

len = 100 # height and width of the grid
permeability = rand(Float32, len, len)

backend = AutoMooncake(; config=nothing)
prep = prepare_gradient(objective, backend, permeability);  # this is SLOW
@time val, grad = value_and_gradient(objective, prep, backend, permeability)  # this is FAST
```
